### PR TITLE
some updates to the GPU package

### LINKED ISSUE #939: GPU Package Segfault
## Summary
I'm getting a segfault when trying to use the GPU package with `/bench/in.lj` or other input scripts. I've reproduced this on two different machines with different types of GPUs.

## Type of Issue
Bug report

## LAMMPS Version (Bug Report)
Latest master = 11 May 2018

## Expected Behavior (Bug Report)
No segfault.

## Actual Behavior (Bug Report)
```
Program received signal SIGSEGV, Segmentation fault.
0x000010000133f2bc in __strlen_power7 () from /lib64/libc.so.6
(cuda-gdb) where
#0  0x000010000133f2bc in __strlen_power7 () from /lib64/libc.so.6
#1  0x00000000104ad6ec in LAMMPS_AL::Device<float, double>::init_device(ompi_communicator_t*, ompi_communicator_t*, int, int, int, double, int, int, double, char*, int) ()
#2  0x0000000000000001 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```
## Steps to Reproduce (Bug Report)

```
cd /bench
~/lammps/src/lmp_gpu -in in.lj -sf gpu -pk gpu 1
```

### PR DESCRIPTION
**Pull request by ndtrung81**
## Purpose

- Fixed a bug with MPI_Allgather when used for node_names dynamically allocated in lal_device.cpp
- Added pair lj/expand/coul/long to USER-MISC and its /gpu version
- Added pair lj/cut/dipole/long/gpu
- Added example Makefile's (Makefile.linux_multi) for lib/gpu to build libgpu.a with multiple architectures

## Author(s)

Trung Nguyen (Northwestern)

## Backward Compatibility

The changes should not affect other parts of LAMMPS.

## Implementation Notes

The new pair styles and their gpu version were tested by minimal input decks derived from examples/dipole.

## Post Submission Checklist

_Please check the fields below as they are completed_
- [x] The feature or features in this pull request is complete
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] One or more example input decks are included
- [x] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

Fixes #939

### CODE REVIEW

**Comment by junghans** [comment]
@rbberger can you test this in CMake?

**Comment by rbberger** [comment]
@ndtrung81 I'm adding a small fix to the CMake build of this. We had this bug for some time.

**Comment by ndtrung81** [comment]
@rbberger Thanks! I didn't notice that.

File: lib/gpu/Nvidia.makefile_multi [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+$(OBJ_DIR):
+	mkdir -p $@
+
+$(OBJ_DIR)/cudpp.o: cudpp_mini/cudpp.cpp
+	$(CUDR) -o $@ -c cudpp_mini/cudpp.cpp -Icudpp_mini
+
+$(OBJ_DIR)/cudpp_plan.o: cudpp_mini/cudpp_plan.cpp
+	$(CUDR) -o $@ -c cudpp_mini/cudpp_plan.cpp -Icudpp_mini
+
+$(OBJ_DIR)/cudpp_maximal_launch.o: cudpp_mini/cudpp_maximal_launch.cpp
```
Comment (junghans) [inline lib/gpu/Nvidia.makefile_multi]: Can we use a pattern rule?

**Comment by rbberger** [comment]
@junghans we also have other problems with CMake. I tested yesterday on Ubuntu, CUDA 9 and Volta and `$<$<>>` doesn't expand properly, causes an NVCC compilation error and even after manually fixing it the final binary segfaults. Doesn't seem to be related to this PR. I'm looking into it.

**Comment by rbberger** [comment]
ok, correction. The segfault disappears with this PR. It was the `MPI_Allgather` bug. I also can't reproduce the CMake issue on Arch with CUDA 9.1 and the TITAN X, so it must be something about the Ubuntu setup.

**Comment by rbberger** [comment]
@junghans looks like this is only an issue with cmake 3.5.1 on Ubuntu. Installing 3.11.2 fixes the issue of not expanding the conditional compilation options passed to `cuda_compile`.

File: lib/gpu/lal_device.cpp [hunk 2]
```diff
@@ -80,7 +80,7 @@ int DeviceT::init_device(MPI_Comm world, MPI_Comm replica, const int first_gpu,
   char node_name[MPI_MAX_PROCESSOR_NAME];
   char *node_names = new char[MPI_MAX_PROCESSOR_NAME*_world_size];
   MPI_Get_processor_name(node_name,&name_length);
-  MPI_Allgather(&node_name,MPI_MAX_PROCESSOR_NAME,MPI_CHAR,&node_names,
+  MPI_Allgather(&node_name,MPI_MAX_PROCESSOR_NAME,MPI_CHAR,&node_names[0],
```
Comment (ndtrung81) [inline lib/gpu/lal_device.cpp]: @stanmoore1 I think the segfault reported in #939 may be resolved with this change in lal_device.cpp. Can you test this for me? Thanks!

Comment (stanmoore1) [reply lib/gpu/lal_device.cpp]: Yes, confirmed. Thanks

**Comment by stanmoore1** [comment]
Fixed #939.

File: lib/gpu/lal_device.cpp (hunk 2 above)
Comment (akohlmey) [reply lib/gpu/lal_device.cpp]: That one was introduced by me when we removed the non-standard variable length arrays. Sorry about that.

**Comment by ndtrung81** [comment]
@akohlmey I have made the changes to the source files in lib/gpu and src/GPU.

**Comment by stanmoore1** [comment]
@ndtrung81 can you also update the header files for `pair lj/expand/coul/long` in `USER-MISC`? Thanks
@akohlmey I don't see where `<cinttypes>` is included

**Comment by ndtrung81** [comment]
@stanmoore1 sure, just did in commit 3c781af

**Comment by stanmoore1** [comment]
@akohlmey the only place I see `<cinttypes>` used is in Kokkos:

```
grep -I -r cinttypes *
lib/kokkos/core/src/impl/Kokkos_Profiling_Interface.hpp:#include <cinttypes>
lib/kokkos/core/unit_test/TestTeamVector.hpp:#include <cinttypes>
src/dump_dcd.cpp:#include <inttypes.h> // <cinttypes> requires C++-11
src/lmptype.h:#include <inttypes.h> // <cinttypes> requires C++-11
```

So I'm assigning this to @sjplimp to be merged.

**Pull request merged**